x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java

```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+                    matcher.addToCompleted(sequence);
+                }
+            }
+        }
+        checkMissingEvents(next, listener);
+    }
+
+    private boolean biggerThanMaxSpan(Timestamp from, Timestamp to) {
+        if (from == null || to == null) {
+            return true;
+        }
+        return to.instant().toEpochMilli() - from.instant().toEpochMilli() >= matcher.maxSpanInNanos() / 1_000_000;
+    }
+
+    private List<SearchRequest> prepareQueryForMissingEvents(List<Sequence> toCheck) {
+        List<SearchRequest> result = new ArrayList<>();
+        for (Sequence sequence : toCheck) {
+            boolean leading = true;
+            for (int i = 0; i < criteria.size(); i++) {
+                SequenceCriterion criterion = criteria.get(i);
+                if (criterion.missing()) {
+                    BoxedQueryRequest r = criterion.queryRequest();
+                    RangeQueryBuilder range = new RangeQueryBuilder(r.timestampField()).timeZone("UTC").format("epoch_millis");
+                    SearchSourceBuilder builder = copySource(r.searchSource());
+                    if (leading) {
+                        builder.sorts().clear();
+                        builder.sort(r.timestampField(), SortOrder.DESC);
+                        range.lt(sequence.startOrdinal().timestamp());
+                    } else if (trailing(i)) {
+                        builder.sorts().clear();
+                        builder.sort(r.timestampField(), SortOrder.ASC);
+                        range.gt(sequence.ordinal().timestamp());
+                    } else {
+                        range.lt(sequence.matchAt(matcher.nextPositiveStage(i)).ordinal().timestamp());
+                        range.gt(sequence.matchAt(matcher.previousPositiveStage(i)).ordinal().timestamp());
+                        builder.sort(r.timestampField(), SortOrder.ASC);
+                    }
+                    addKeyFilter(i, sequence, builder);
+                    RuntimeUtils.addFilter(range, builder);
+                    result.add(RuntimeUtils.prepareRequest(builder.size(1).trackTotalHits(false), false, Strings.EMPTY_ARRAY));
+                } else {
+                    leading = false;
+                }
+            }
+        }
+        return result;
+    }
+
+    private void addKeyFilter(int stage, Sequence sequence, SearchSourceBuilder builder) {
+        List<Attribute> keys = listOfKeys.get(stage);
+        if (keys.isEmpty()) {
+            return;
+        }
+        for (int i = 0; i < keys.size(); i++) {
+            Attribute k = keys.get(i);
+            addFilter(new TermQueryBuilder(k.qualifiedName(), sequence.key().asList().get(i)), builder);
+        }
+    }
+
+    private boolean trailing(int i) {
+        return matcher.nextPositiveStage(i - 1) < 0;
+    }
+
+    /*
+     * Not a great way of getting a copy of a SearchSourceBuilder
+     */
+    private static SearchSourceBuilder copySource(SearchSourceBuilder source) {
```

Comment:
For this method, please create an `ExecutionUtils` class under `org.elasticsearch.xpack.eql.execution` package and add it there (if you have other ideas/options please let me know). Replace the call in SampleQueryRequest, as well.